### PR TITLE
Pitch: Changed range from 0-100 to 1-101 to avoid ZeroDivisionError

### PR DESCRIPTION
**Pull request by ShubhamJain7**
### Link to issue number:
https://github.com/nvaccess/nvda/issues/10940

### Summary of the issue:
When the pitch is set to 0, a `ZeroDivisionError` occurs when trying to speak capital letters. This is because the `baseProsodyCommand.multiplier` property used to calculate the new value of pitch, to be used when speaking capital letters, relies on a division operation with the original pitch (`defaultVal`) in the denominator.

### Description of how this pull request fixes the issue:
The first solution I tried, was to use `baseProsodyCommand.offset` instead of the `baseProsodyCommand.multiplier` to perform all calculations but this was infeasible since the values for the prosody attribute Rate must be a relative non-negative-percentage value (refer [here](https://www.w3.org/TR/2010/REC-speech-synthesis11-20100907/#number_values)) in the SSML. Using `baseProsodyCommand.offset` to calculate this value would replicate the logic in `baseProsodyCommand.multiplier` and thus result in the same error. Another issue with this solution is that a lot of other code was written to be used with the `baseProsodyCommand.multiplier`.

The second solution I tried, was to catch the `ZeroDivisionError` in `baseProsodyCommand.multiplier`  and raise it again to be handled by any code that calls it. Although the exception had to be handled a little differently for each case, the most common method was to set the multiplier to a fixed value (0 and multiplier values for other possible values of `baseProsodyCommand.defaultValue` or the pitch value stored in the config). The pitch value stored in the config is multiplied with `baseProsodyCommand.multiplier * 100` to calculate the new pitch to be used for speaking capital letters. Since the pitch value stored in the config is still 0 in this case, the result is 0 too and hence there is no pitch change.

The third solution was to change the range of pitch from 0-100 to 1-101, this way, a pitch value of 0 is avoided entirely. It was necessary to use a range of 1-101 and not 1-100 because it caused scaling issues and the value of 99 was omitted. This is apparent in speech settings GUI as the pitch slider announces all values except 99 when incrementing the slider by 1 from 0 to 100. An artefact of this solution is that the value for pitch stored in the config is 1 more than what the user sets in the speech settings GUI. For example, if the pitch is set to 0, the value stored in the config is 1 and if the pitch is set to 100, the value stored in the config is 101. Since pitch values of 0 are now invalid,
this change also required a profile upgrade step so the program won't crash for users who have pitch set to 0. 
This change of range was not applied for the other prosody attributes (Rate and Volume) since they do not cause errors when they are set to zero. However, as a precaution, a warning is logged for the two prosody settings if set to zero and created using non-default values of offset or multiplier.
Although not perfect, it was found that this solution had the least negative user impact and code change.

### Testing performed:

1. Set Pitch=0 CapPitchChange percentage=0
2. Open notepad and type lowercase and uppercase letters and check if pitch change could be heard
3. Repeat the previous steps with negative and positive values of CapPitchChange percentage (-100, -50, 50, 100)
4. Repeat the previous steps with different values of pitch (25, 50, 75,100)
5. Repeat the previous steps for the other prosody attributes: Rate and Volume 
6. Repeat the above test with different synth drivers: espeak, oneCore and sapi5

### Known issues with pull request:
None

### Change log entry:
not necessary

Section: New features, Changes, Bug fixes

### CODE REVIEW

**Comment by JulienCochuyt** [comment]
@ShubhamJain7 wrote:
> The third solution was to change the range of pitch from 0-100 to 1-101. It was necessary to use a range of 1-101 and not 1-100 because it caused scaling issues and the value of 99 was omitted. This change also required a profile upgrade step so the program won't crash for users who have pitch set to 0.
The profile upgrade could have been omitted altogether if the pitch value was incremented by 1 only before being used - or maybe is it read directly from config much to often?

Also, despite we all prefer well linted code, beware that linting code unrelated to your change makes it more difficult to review now and blame in the future.

**Comment by ShubhamJain7** [comment]
@JulienCochuyt 

> or maybe is it read directly from config much too often

It is and so it would be tedious to increment it by 1 everywhere. 

>beware that linting code unrelated to your change makes it more difficult to review now and blame in the future.

This is why I left them out in my first commit, but it caused AppVeyor's lint test to fail so I went back and fixed them.

**Comment by josephsl** [comment]
Hi, eh? Did you run scons lint on your local source code copy before committing? If yes yet Appveyor gives lint errors, it might be something we should discuss on a different issue. Thanks.

**Comment by ShubhamJain7** [comment]
@josephsl
>Did you run scons lint on your local source code copy before committing

I did and these errors showed up there too, but I decided not to do anything about it since the errors were unrelated to the changes made by me.

**Comment by josephsl** [comment]
Hi, I see (try running scons lint base=origin/master so only changes you have made are inspected). Thanks.

**Comment by ShubhamJain7** [comment]
I'll keep that in mind the next time.
Thanks!

**Comment by leonardder** [comment]
I'm not sure whether I'm very happy with this change, it feels a bit hacky to me to have a 1/101 range.

> The second solution I tried, was to catch the `ZeroDivisionError` and handle it by setting the multiplier to zero or a non-zero value. This didn't work perfectly since the pitch was still zero and multiplying anything with zero results in zero. So there is no change in pitch when speaking capital letters.

Yet I think that this behavior was the behavior before we moved to Python 3.

Have you considered changing the multiplier calculation like this?

```
	@property
	def multiplier(self):
		...
		# Calculate multiplier from default value and offset.
		defaultVal = self.defaultValue
		newVal = defaultVal + self._offset
		return newVal / max(1, defaultVal)

```

note that `float` is no longer needed.

**Comment by ShubhamJain7** [comment]
@leonardder 
I tried something similar to what you are suggesting (refer to the second solution in the PR description) however it did not work for the same reason my solution did not. Since the value of pitch in the config is still 0, any percentage increase or decrease will still be 0 and thus no pitch change is heard.
But yes, `float` is no longer necessary when using the `/` in python 3.

**Comment by leonardder** [comment]
The important part is the use of max, that ensures that the pitch 
couldn't go below 1 while calculating the multiplier. Are you saying 
this does not work?

**Comment by ShubhamJain7** [comment]
@leonardder 
Yes. 
Your solution certainly prevents the `ZeroDivisionError` from occurring by ensuring that the denominator is 1 in case `defaultVal` is 0. However, since the pitch stored in the config is still 0, any multiplier value used with it will still result in 0. This means that no pitch change can be heard when the user sets the pitch to 0 but it can be heard for other values of pitch. This is not the behaviour a user would expect.

**Comment by leonardder** [comment]
I see.

I believe the pitch as stored in the config is always used by using the defaultValue property on the BaseProsodyCommand within the scope of handling commands.

How about overriding that property on the PitchCommand class, like this:

```
	@property
	def defaultValue(self):
		return max(super().defaultValue, 1)
```

**Comment by feerrenrut** [comment]
@leonardder This approach would be quite deceptive in in the cases where this is used with a config item that is valid at 0 along with the `offset` property. 

Given that we are quite tied to using percentage based variation for altering synth pitch, and 0 can not work with this, I personally think this is most clearly defined via the config. The user doesn't see these values, the GUI still reports 101 values (0-100 inclusive). It's a frustrating problem that is made worse by the duel responsibility of the `BaseProsody` command. 

I'm quite against special-casing defatultValue=0 in the BaseProdsody command code, this could cause hard to track down problems in the future.

I went through various alternatives with @ShubhamJain7, this seems like the best option to me. I'll hold off on merging for a week in-case anyone can make a better suggestion. 

One thing that isn't mentioned here is whether this error could potentially occur for other prosody commands, I suspect it could.

File: source/speech/commands.py [hunk 1]
```diff
@@ -229,21 +239,41 @@ def __repr__(self):
...
 	"""
 	settingName = "volume"
 
+	def __init__(self, offset=0, multiplier=1):
+		if not (offset == 0 and multiplier == 1):
+			log.warning(
+				f'If defaultVal of Rate attribute is 0, calling the multiplier property '
+				f'will result in a ZeroDivisionError.'
+			)
+		super().__init__(offset, multiplier)
```
Comment (feerrenrut) [inline source/speech/commands.py:260]: I don't this warning is very helpful at the moment. The condition should also check if defaultVal is 0. Better yet would be to consider the implications of taking the same approach for Volume and Rate.

**Review by feerrenrut — CHANGES_REQUESTED** [review]
Another approach might be to override the defaultValue property on PitchCommand, RateCommand, and VolumeCommand and remap values to avoid zero.

**Comment by leonardder** [comment]
Hmm, I'm still tempted to find a 1 -- 101 range very ugly, but I understand your concern.
@jcsteh as a speech refactor author, Could you think of an alternative?

**Comment by jcsteh** [comment]
> Your solution certainly prevents the `ZeroDivisionError` from occurring by ensuring that the denominator is 1 in case `defaultVal` is 0. However, since the pitch stored in the config is still 0, any multiplier value used with it will still result in 0. This means that no pitch change can be heard when the user sets the pitch to 0 but it can be heard for other values of pitch. This is not the behaviour a user would expect.

I don't follow this. @leonardder's proposed code in https://github.com/nvaccess/nvda/pull/11178#issuecomment-631613821 will mean that the multiplier returned will be non-0. For example, if the current pitch is 0 and the offset is 30, it will return 30, since 30 / 1 = 30. If you literally multiply 0 * 30, then sure, you'll get 0. However, we're passing the multiplier directly to the synth, and the synth shouldn't do that internally: even though the default pitch is 0, that is still relative to some base frequency.

Even if I'm wrong and synths do handle this incorrectly, the proposed fix in this PR won't help, unless there's an unintentional side effect. Yes, you're dealing with values from 1 to 101, but when the default value is passed to the synth, it will be scaled back to the synth's values; e.g. for eSpeak, this is 0 to 99. So, user configured pitch 1 will end up telling eSpeak to set pitch to 0. At that point, you've got exactly the same problem.

**Comment by jcsteh** [comment]
> If you literally multiply 0 * 30, then sure, you'll get 0. However, we're passing the multiplier directly to the synth, and the synth shouldn't do that internally: even though the default pitch is 0, that is still relative to some base frequency.

Apparently, eSpeak at least does handle this incorrectly.

1. Set pitch to 0.
2. Run this in the Python console:
    `import synthDrivers._espeak; wx.CallLater(3000, synthDrivers._espeak.speak, '<prosody pitch="3000%">foo</prosody>')`
3. Wait 3 seconds.
    - Expected: eSpeak should say "foo" at pitch 30.
    - Actual: eSpeak says "foo" at pitch 0.

However, this means that the solution proposed here isn't going to fix the issue because of scaling. The only way to avoid this is to never allow a 0 value to be passed to synths which use multipliers and handle this incorrectly. For eSpeak, that means setting MIN_PITCH to 1 instead of 0.

**Comment by leonardder** [comment]
Given the complexity of this, I think it would be helpful to split this pr into two:

1. One pr that restores the Python 2 bhavior (i.e. speech should no longer stall because of zero division errors)
2. One pr that fixes capital pitch change when pithc is set to 0%

I understand that it makes sense to do this all in one pr, but whereas a zero division error fix could make it into 2020.2, I doubt that we could come up with something stable in time for the other issue.

**Pull request closed without merging**